Fix #383 : Limit the visible length of the probe descriptions

Sylvia23 opened the pull request:
If a probe has long caption, then it doesn't look good if shown completely in the UI. This PR has resolved this issue. Knowledge of HTML, CSS(especially styling `text-overflow: ellipsis;` ) and core JavaScript is required.

georgf reviewed (commented):
This looks good visually!
I'm mostly wondering whether we need the width limit here, the rest are smaller style issues etc.

Sylvia23 committed:
Fix #383 : Limit the visible length of the probe descriptions

Sylvia23 commented:
Hi @georgf , As of now, I have commited this PR without the changes you suggested.
I will change the code as per you suggestions and then commit again.
Thanks!

georgf commented:
Sounds good, thanks.

Sylvia23 commented:
@georgf , Unfortunately, I couldn't see the changes that you suggested. It seems that they have gone. I can only see the comments in the mail but not the references. So, I am unable to figure out the suggested changes.
Can you please review it again. I am extremely sorry for the inconvenience.
Thanks!

georgf commented:
No worries, will do!

georgf commented on new-pipeline/evo.html:
```diff
@@ -60,7 +60,7 @@ <h2 style="font-size: 18px; margin-bottom: 0">
                     <select id="filter-process-type" class="multiselect" title="Process type" data-all-selected="any process type"></select>
                 </h2>
-                <h3 id="measure-description" style="font-size: 14px; margin: 0 3px"></h3>
+                <h3 id="measure-description"></h3>
```
Nice cleanup!

Sylvia23 replied:
Thanks :D

georgf commented on new-pipeline/src/dashboards.js:
```diff
@@ -1077,7 +1077,13 @@ function getDescriptionWithLink(metric, channel, description) {
   if (description && (description.length > 0)) {
-    descr = description;
+    if (description.length > 85) {
+      descr = description.substring(0, 85);
```
Do we need this JavaScript change?
It would be great if we can do this just in CSS, so the dashboard can e.g. automatically show longer descriptions for larger screens.

Sylvia23 replied:
Actually, I did the same initially, but the "More details" option doesn't come without this JavaScript snippet. While,  CSS decreases the length of the long description.

Sylvia23 replied:
@georgf , Now, I got what you exactly want. I am working on it now. Hopefully, it will be ready in sometime. Actually, It requires very involved code change. I am fixing the bugs that I am getting.
Thanks!

georgf commented on new-pipeline/style/dashboards.css:
```diff
@@ -126,10 +126,15 @@
+#dist-caption, #measure-description {
+  margin-left: 2em;
+  white-space: nowrap;
+  width : 55em;
```
Do we need to limit the width?
It would be great if we could just use the full available page width for the description text.

Sylvia23 replied:
Okay ! Will do !

georgf reviewed (commented):
Visually this looks good.
It would be nice though if we can use the full available width for the description text.
Does this work?

Sylvia23 commented:
@georgf , I have removed the width. Now, we can use the full width.
Can you review the PR.
Thanks!

Sylvia23 commented:
@georgf , Can you please review this PR. Thanks!

Sylvia23 committed:
Fix #383 : Limit the visible length of the probe descriptions

georgf commented on new-pipeline/src/dist.js:
```diff
@@ -886,9 +886,12 @@ function displayHistograms(histogramsList, dates, useTable, cumulative, trim) {
 
     $('#dist-caption').html(desc);
+    $('#caption-dots').html("...");
+    $('#dist-caption-link').html(link);
```
I think `text-overflow: ellipsis;` should already take care of adding the "...".

Sylvia23 replied:
@georgf , Okay, I will look into it and report ASAP. Thanks!

georgf commented on new-pipeline/src/dashboards.js:
```diff
@@ -1085,6 +1084,16 @@ function getDescriptionWithLink(metric, channel, description) {
+function getDescriptionLink(metric, channel, description) {
+  var metricUrl = buildDictionaryURL(metric, channel, description);
+
+  var div = $("<div>", {
```
Is this `div` element used somewhere?

Sylvia23 replied:
No, I will remove it. Thanks!

georgf reviewed (commented):
Sorry for missing this before, i didn't see it in my Github notifications.

I think we don't need JavaScript to add the ellipsis, that should be possible with just the `text-overflow` CSS.

Testing this locally, this doesn't cut off to overflow currently (see screenshot).
We should make this first, work out how to make the overflow work from CSS only.
There are some other comments below, but they are not so important right now.

![screenshot-2018-3-2 measurement dashboard 1](https://user-images.githubusercontent.com/407696/36912178-3ca33646-1e46-11e8-8268-681dec68f7b1.png)

Sylvia23 commented:
@georgf , I have used CSS for ellipsis. I removed the JS part. This shortening of caption takes place by CSS `text-overflow : ellipsis` only now. JS is used just to dispaly the caption. Now, it will work for screen of any size which was the concern by using JS.
Please have a look!
Thanks!

Sylvia23 commented:
I just saw , the `overflow` is not working propperly. I will fix it now.
Thanks!

Sylvia23 commented:
@georgf , I have fixed it. Can you please review the PR. Thanks!

georgf commented on new-pipeline/dist.html:
```diff
@@ -55,7 +55,11 @@ <h2 style="font-size: 18px; margin-bottom: 0">
-                  <figcaption id="dist-caption"></figcaption>
+                  <figcaption id="dist-caption" class="overflow ellipsis"></figcaption>
+                  <div class="caption-div">
+                  <figcaption id="caption-dots" >.....</figcaption>
```
I think this is not working right yet. The "..." should be added automatically through the "text-overflow: ellipsis;" command.
Let's maybe try this isolated without all the existing code around first. I think [this is](https://jsfiddle.net/exzLLez0/11/) roughly the structure that we want here (not pretty, but shows the idea).

Can we apply this idea in this dashboard?

Sylvia23 replied:
@georgf ,  I had to add "..." separately because If we put the content(caption) directly into the `HTML` tag `<figcaption></figcaption>`, then `"..."` comes automatically.
But if the caption is rendered using JavaScript, then this `"..."` doesn't come.
Otherwise, the shortening of caption is taking place as per the desired UI.

Please see the screenshot attached.

![screenshot from 2018-03-06 02-11-45](https://user-images.githubusercontent.com/23060217/36998867-6293ff02-20e4-11e8-9fdf-1cfbfd8ec933.png)
![screenshot from 2018-03-06 02-11-52](https://user-images.githubusercontent.com/23060217/36998868-62c77d1e-20e4-11e8-9c09-e32a9888883f.png)

Any advice on what and where to get the changes done is most welcome! 
Thanks!

georgf replied:
I think the ellipsis in the screenshot looks fine. The only issues is that the "more details" button is on a new line.
Am i missing some problem here? :)

Sylvia23 replied:
The `ellipsis` looks fine in the screen-shot because I have put an example caption in the `HTML`tag directly.  Please see the code in the screen-shot.

![screenshot from 2018-03-06 02-11-45](https://user-images.githubusercontent.com/23060217/37000442-6c97e2ca-20e9-11e8-88f0-e7a4ad91f310.png)

This code shows up like : 

![screenshot from 2018-03-06 02-11-52](https://user-images.githubusercontent.com/23060217/37000758-596280ce-20ea-11e8-9a28-f52a092e6690.png)

The problem comes when the caption is rendered using JavaScript in the runtime from `getDescription` function here [https://github.com/mozilla/telemetry-dashboard/blob/gh-pages/new-pipeline/src/dashboards.js#L1059](url) .
In this case, the caption does shorten up like what a normal `ellipsis` does except `"..."` doesn't come. Hence, I had to add `"..."` separately.

![screenshot from 2018-03-06 02-57-23](https://user-images.githubusercontent.com/23060217/37000767-6112a524-20ea-11e8-9541-f2e1c820da3b.png)

This shows up like :

![screenshot from 2018-03-06 03-00-37](https://user-images.githubusercontent.com/23060217/37000779-6d8bc132-20ea-11e8-9625-f370f97df061.png)

Does this sound clear?

Sylvia23 replied:
@georgf , I have found a fix for my problem. I will report back in a while. Thanks!

Sylvia23 commented:
@georgf , Now, the UI looks like : 

For long captions: 
![screenshot from 2018-03-06 05-39-20](https://user-images.githubusercontent.com/23060217/37006879-871ac934-2100-11e8-89e3-68b7e3b4ca6a.png)

For short captions:
![screenshot from 2018-03-06 05-40-50](https://user-images.githubusercontent.com/23060217/37006932-c28539b4-2100-11e8-8bfe-1afb6a455758.png)

One thing is, the caption and "More Details" can't be aligned in one line using `CSS` because `ellipsis` and `inline-block` don't work together as far as I know. One way is to fix the `width` of caption (e.g 80%) but then it will create problem with short captions. 
Any advice with the implementation is most welcomed. Thanks !

georgf commented:
I did a little research and found that [this jsfiddle](https://jsfiddle.net/y7404myo/3/) seems to do what we want here: with small window widths the text on the left overflows with an ellipsis, but the button on the right stays unaffected.
Can you try that approach?

Sylvia23 commented:
@georgf , Yes, `display : flex` worked for me rather than `display : inline-block`. Thanks for the guidance.
It is good for review now.

georgf commented:
Great, i will look over it tomorrow!

Can you publish your dashboard through Github pages and comment with the live link here? That makes it much easier to check how the changes look.
You can enable this in your forked repository under "settings", "Github pages".

Sylvia23 commented:
@georgf , I have already done this.
Here is the link : [https://sylvia23.github.io/telemetry-dashboard/new-pipeline/dist.html#!cumulative=0&end_date=2018-03-05&keys=__none__!__none__!__none__&max_channel_version=nightly%252F60&measure=SCALARS_BROWSER.SESSION.RESTORE.BROWSER_STARTUP_PAGE&min_channel_version=nightly%252F57&processType=*&product=Firefox&sanitize=1&sort_keys=submissions&start_date=2018-01-22&table=0&trim=1&use_submission_date=0](url)

Sylvia23 commented:
@georgf , I have already done it. Here is the link : [https://sylvia23.github.io/telemetry-dashboard/](url).
Are you able to see this?

georgf commented:
Yes, i see that, cheers!

One problem i notice on that page:
When i load it it looks fine. But when i reduce the window width, suddenly the description is shown multiple times:
![screen shot 2018-03-07 at 14 20 21](https://user-images.githubusercontent.com/407696/37094348-ed24999a-2212-11e8-86fa-d2823f833048.png)

Checking this in the inspector, it looks like the text is actually added to the page multiple times:
![screen shot 2018-03-07 at 14 22 36](https://user-images.githubusercontent.com/407696/37094406-16af7b40-2213-11e8-852f-6b8e0938e5e3.png)

Sylvia23 commented:
@georgf , Should we do like , we reload the page whenever the user resize the window to fix this issue!?
Please suggest. Thanks!

georgf commented:
@Sylvia23 I think the problem is in the JavaScript changes here. It looks like we are appending the text repeatedly instead of replacing it.
This is a new problem, so it is probably a problem in the JavaScript that you changed.
Does that help tracking the problem down?

Sylvia23 commented:
@georgf , I have fixed it. Actually I changed the `$('dist-caption').html(desc)` to `$('dist-caption').innerHTML += desc` due to which it was appending the caption everytime the function gets called. 
Now, it's good to go. Please review. Thanks!

georgf commented on new-pipeline/dist.html:
```diff
@@ -55,7 +55,10 @@ <h2 style="font-size: 18px; margin-bottom: 0">
                 <figure class="col-md-12">
-                  <figcaption id="dist-caption"></figcaption>
+                <div class="caption-div">
+                  <figcaption id="dist-caption" class="overflow ellipsis"></figcaption>
```
A small request: let's call this `dist-caption-text`, to better tell the two elements apart.

Sylvia23 replied:
Sure!

georgf commented on new-pipeline/evo.html:
```diff
@@ -54,6 +54,13 @@ <h2 style="font-size: 18px; margin-bottom: 0">
                 </h2>
+                <hr></hr>
+                <div class="caption-div">
+                <figcaption id="evo-caption" class="overflow ellipsis"></figcaption>
```
Likewise, lets use `evo-caption`. This and the following element are missing indentation (should be 2 spaces per level).

Sylvia23 replied:
Here, you mean `evo-caption-text` Right?

georgf commented on new-pipeline/evo.html:
```diff
@@ -54,6 +54,13 @@ <h2 style="font-size: 18px; margin-bottom: 0">
+                <figcaption id="evo-caption" class="overflow ellipsis"></figcaption>
+                <figcaption id="evo-caption-link"></figcaption>
+                </div> 
+                <h2 style="font-size: 18px; margin-bottom: 0">Showing graphs for <select id="selected-key" class="multiselect" title="Selected key"></select></h2>
```
This is duplicated, you can see [here](https://bit.ly/2FBu2ac) how it shows up two times.

Sylvia23 replied:
Yeah, I see.

georgf commented on new-pipeline/src/dist.js:
```diff
@@ -874,9 +874,11 @@ function displayHistograms(histogramsList, dates, useTable, cumulative, trim) {
-
+    var desc = getDescription(metric, channel, description)
+    var link = getDescriptionLink(metric, channel, description)
+    var div = document.getElementById('dist-caption');
```
This is more clearly structured now, nice!

Sylvia23 replied:
Yeah! 😀

georgf reviewed (commented):
This looks good now, cheers!
I just have some comments below on smaller cleanup, then this should be done.

We should improve the text styling later, but that is not important here. #406 will address that part.

Sylvia23 commented:
@georgf commited the changes requested. Ready for final review. Thanks!

georgf commented on new-pipeline/dist.html:
```diff
@@ -55,7 +55,10 @@ <h2 style="font-size: 18px; margin-bottom: 0">
                 <hr></hr>
                 <figure class="col-md-12">
-                  <figcaption id="dist-caption"></figcaption>
+                <div class="caption-div">
```
This `div` is missing one indentation level. Then the `figcaption` need to move along accordingly.

georgf commented on new-pipeline/dist.html:
```diff
@@ -55,7 +55,10 @@ <h2 style="font-size: 18px; margin-bottom: 0">
+                <div class="caption-div">
+                  <figcaption id="dist-caption-text" class="overflow ellipsis"></figcaption>
+                  <figcaption id="dist-caption-link"></figcaption>
+                </div></br></br>
```
I think we should not use `br` here. Instead lets use `margin-bottom` or similar.

georgf commented on new-pipeline/src/dist.js:
```diff
@@ -874,11 +874,13 @@ function displayHistograms(histogramsList, dates, useTable, cumulative, trim) {
-
-    $('#dist-caption').html(desc);
+    var desc = getDescription(metric, channel, description)
+    var link = getDescriptionLink(metric, channel, description)
```
These lines are missing a semicolon at the end.

georgf commented on new-pipeline/src/evo.js:
```diff
@@ -324,11 +324,12 @@ $(function () {
-              var description = getDescriptionWithLink(metric, channel, evolutionDescription)
-
+              var description = getDescription(metric, channel, evolutionDescription)
+              var link = getDescriptionLink(metric, channel, description)
```
These lines are also missing semicolons.

georgf reviewed (commented):
Thanks! There are still a few small things i noticed, but visually this is done :)

Sylvia23 commented:
@georgf pushed the changes requested. Thanks!

georgf commented on new-pipeline/dist.html:
```diff
@@ -55,7 +55,10 @@ <h2 style="font-size: 18px; margin-bottom: 0">
+	                  <figcaption id="dist-caption-text" class="overflow ellipsis"></figcaption>
+	                  <figcaption id="dist-caption-link"></figcaption>
+	                </div>
                     <div class="col-md-9" id="plots">
```
The indentation is still a bit off here.
The goal would be to have `dist-caption` on the same indentation level as `plots` (as they are at the same level in the hierarchy here).

Sylvia23 replied:
@georgf , I put `dist-caption` on the same level as that of `plots` and again pushed. But it still shows it like this, they are on the same level in my code on my local machine. Can you please review and tell the probable reasons. Thanks!

georgf replied:
Are you mixing tabs and spaces here?
Most code editors have an option to show these with extra characters.
In the code in this repository, we only use spaces for indentation, not tabs.

Sylvia23 replied:
@georgf , This is how the code looks like in my editor. 

![screenshot from 2018-03-10 13-17-39](https://user-images.githubusercontent.com/23060217/37239633-70bd9de8-2465-11e8-9603-e54a7758250d.png)

Please have a look. Thanks!

georgf replied:
If you look in the ["files changed" view](https://github.com/mozilla/telemetry-dashboard/pull/418/files) you can see that it looks different.

To fix this, you can check for an option to show tabs & spaces in your editor, so you can see this.
More generally you can check for an option in your editor that let's you choose to use spaces instead of tabs for indentation.

Sylvia23 replied:
@georgf , Yeah I can see the difference when I push the branch on GitHub. But it is not visible in my editor. I already have `Indent using spaces` setting in my editor. I will try again and hope to fix it. Thanks!

georgf reviewed (commented):
Thanks - one last small change needed, then we're good!

Sylvia23 commented:
@georgf , I think it's fixed now. We are now good to go. Thank you so much!

Sylvia23 commented:
@georgf , Can you please review this whenever you are free. Thanks!

chutten commented on new-pipeline/dist.html:
```diff
@@ -55,8 +55,11 @@ <h2 style="font-size: 18px; margin-bottom: 0">
-                  <figcaption id="dist-caption"></figcaption>
-                    <div class="col-md-9" id="plots">
+                        <div class="caption-div">
+                            <figcaption id="dist-caption-text" class="overflow ellipsis"></figcaption>
```
It doesn't matter to browsers, but it might interest you to know that the spec expects `figcaption` elements to be either the first or last child (or one of each) of a `figure` element. [ref](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/figure)

chutten commented on new-pipeline/src/dist.js:
```diff
@@ -874,11 +874,13 @@ function displayHistograms(histogramsList, dates, useTable, cumulative, trim) {
+    $('#dist-caption-link').html(link);
   } else {
-    $('#dist-caption').text(""); // Clear the histogram caption
+    $('#dist-caption-text').text(""); // Clear the histogram caption
```
Should we also clear the link? (it is possible this case is never reached, so it may not matter except for symmetry)

chutten commented on new-pipeline/style/dashboards.css:
```diff
@@ -126,10 +126,27 @@
+  align-items: center;
   text-align: center;
-  margin: 0.1em;
+  margin: 0 auto;
```
`.caption-div` has full width, so automatic horizontal margins like these will always calculate to 0. I think we can omit this.

chutten commented on new-pipeline/evo.html:
```diff
@@ -54,7 +54,11 @@ <h2 style="font-size: 18px; margin-bottom: 0">
-                <h3 id="measure-description" style="font-size: 14px; margin: 0 3px"></h3>
+                <hr></hr>
+                <div class="caption-div">
+                    <figcaption id="evo-caption-text" class="overflow ellipsis"></figcaption>
```
I think the `.ellipsis` class may be holding on from an earlier version of this patch. It looks like `.overflow` takes care of it now, so we can remove it here and in `dist.html`.

Sylvia23 replied:
I removed it from `dist.html` earlier. I could not see it there now. Please correct me if I am missing something.

Sylvia23 replied:
Ohh I got it I guess. You are saying that I should keep the class name as `overflow` rather than `overflow ellipsis`. Right?

chutten replied:
Correct!

chutten commented on new-pipeline/style/dashboards.css:
```diff
@@ -126,10 +126,27 @@
+.overflow {
+    white-space: nowrap !important;
+    overflow: hidden !important;
+    text-overflow: ellipsis !important;
```
`!important` used to be (may still be? It's been a few years since I've asked) only used if absolutely necessary. I think we can omit them here, as no other classes are trying to set `overflow`, `text-overflow`, or `white-space` on elements with `.overflow`

georgf commented on new-pipeline/dist.html:
```diff
@@ -55,8 +55,11 @@ <h2 style="font-size: 18px; margin-bottom: 0">
+                            <figcaption id="dist-caption-text" class="overflow ellipsis"></figcaption>
+                            <figcaption id="dist-caption-link"></figcaption>
+                        </div>
+                        <div class="col-md-9" id="plots">
```
This looks like the spaces are sorted now!
Now we just need to indent `dist-caption` and `col-md-9` one level less, as they're not on the same level as the following `<div class="row">`.

chutten reviewed (commented):
I'm not @georgf but I'm helping out (he's going to be on vacation soon, so I'm picking things up where I can) : )

I have a couple of questions, a couple of comments, and some CSS opinions. At this point I've tested the changes and the behave perfectly well, so these are minor code cleanups. If you feel as

Sylvia23 commented:
@chutten @georgf , I have made the suggested changes. Please have a look. Thanks!

chutten commented on new-pipeline/style/dashboards.css:
```diff
@@ -126,10 +126,26 @@
+}
+
+.overflow {
+    white-space: nowrap;
```
This file uses two-space indents. This selector's block is at four

chutten reviewed (commented):
Argh, I just noticed the CSS indentation. I'm sorry I didn't notice it last time.

chutten commented:
We _really_ need a linter to catch these things on developer machines. The prototype will definitely have one.

Sylvia23 commented:
@chutten , Yeah, linter automatically analyses the code for these errors. It's a kind of must for this.

I have corrected the indentation. Please see. Thanks!

chutten commented on new-pipeline/src/dist.js:
```diff
@@ -874,11 +874,14 @@ function displayHistograms(histogramsList, dates, useTable, cumulative, trim) {
-    $('#dist-caption').html(desc);
+    var desc = getDescription(metric, channel, description);
+    var link = getDescriptionLink(metric, channel, description);
+    var div = document.getElementById('dist-caption-text');
```
What is div used for?

Sylvia23 replied:
We can drop it. I used it initially for something. But now its not needed. I have removed it.

chutten commented on new-pipeline/style/dashboards.css:
```diff
@@ -126,10 +126,26 @@
+}
+
+.overflow {
+	white-space: nowrap;
```
The whitespace appears to still be off. It looks like it's a tab character instead of spaces?

Sylvia23 commented:
@chutten , I have removed the white space. Does they still appear? Please see. Thanks!

chutten reviewed (approved):
Looks good now!
